sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala

```diff
@@ -60,14 +56,132 @@ object GpuScalar extends Arm with Logging {
         case DType.STRING => UTF8String.fromBytes(v.getUTF8)
         case dt: DType if dt.isDecimalType => Decimal(v.getBigDecimal)
-        case t => throw new IllegalStateException(s"Extracting data from a cudf Scalar is not" +
-          s" supported for type $t.")
+        // Extracting data for a list scalar is not supported, until there is a requirement
+        // from expression computations on host side.
+        // For now, it is only used to expand to a list column vector.
+        case t => throw new UnsupportedOperationException(s"Extracting data from a cudf Scalar" +
+          s" is not supported for type $t.")
       }
     } else {
       null
     }
   }
 
+  /**
+   * Resolves a cudf `HostColumnVector.DataType` from a Spark `DataType`.
+   * The returned type will be used by the `ColumnVector.fromXXX` family.
+   */
+  private[rapids] def resolveElementType(dt: DataType): HostColumnVector.DataType = dt match {
+    case ArrayType(elementType, _) =>
+      new HostColumnVector.ListType(true, resolveElementType(elementType))
+    case StructType(fields) =>
+      new HostColumnVector.StructType(true, fields.map(f => resolveElementType(f.dataType)): _*)
+    case other =>
+      new HostColumnVector.BasicType(true, GpuColumnVector.getNonNestedRapidsType(other))
+  }
+
+  /**
+   * The "convertXXXTo" functions are used to convert the data from Catalyst type
+   * to the Java type required by the `ColumnVector.fromXXX` family.
+   *
+   * The data received from the CPU `Literal` has been converted to Catalyst type already.
+   * The mapping from DataType to Catalyst data type can be found in the function
+   * 'validateLiteralValue' in Spark.
+   */
+  /** Converts a decimal, `dec` should not be null. */
+  private def convertDecimalTo(dec: Decimal, dt: DecimalType): Either[Integer, JLong] = {
+    if (dec.scale > dt.scale) {
+      throw new IllegalArgumentException(s"Unexpected decimals rounding.")
+    }
+    if (!dec.changePrecision(dt.precision, dt.scale)) {
+      throw new IllegalArgumentException(s"Cannot change precision to $dt for decimal: $dec")
+    }
+    if (DecimalType.is32BitDecimalType(dt)) {
+      Left(dec.toUnscaledLong.toInt)
+    } else {
+      Right(dec.toUnscaledLong)
+    }
+  }
+
+  /** Converts an element for nested lists */
+  private def convertElementTo(element: Any, elementType: DataType): Any = elementType match {
+    case _ if element == null => null
+    // StructData does not support utf8 string yet, so parses it here instead of calling the
+    //  `convertUTF8StringTo`. Tracked by https://github.com/rapidsai/cudf/issues/8137
+    case StringType => element.asInstanceOf[UTF8String].toString
+    case dt: DecimalType => convertDecimalTo(element.asInstanceOf[Decimal], dt) match {
+      case Left(i) => i
+      case Right(l) => l
+    }
+    case ArrayType(eType, _) =>
+      val data = element.asInstanceOf[ArrayData]
+      data.array.map(convertElementTo(_, eType)).toList.asJava
```

Comment:
is there a possibility that `data.arrray` equals null?

Reply:
It can not be null according to the Spark code of [ArrayData creation](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala#L165) .